File: src/com/cattailsw/nanidroid/CatTailApplication.java
package com.cattailsw.nanidroid;

import android.app.Application;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import com.cattailsw.nanidroid.util.AnalyticsUtils;
import com.cattailsw.nanidroid.util.CrashReporting;

public class CatTailApplication extends Application {

    public final void onCreate() {
	super.onCreate();
	if (isDeviceValidationNoTelemetry()) {
	    AnalyticsUtils.setDeviceValidationNoTelemetry(true);
	    return;
	}
	CrashReporting.initialize(this);
    }

    private boolean isDeviceValidationNoTelemetry() {
	try {
	    ApplicationInfo info = getPackageManager().getApplicationInfo(
		    getPackageName(), PackageManager.GET_META_DATA);
	    Bundle metadata = info.metaData;
	    return metadata != null && metadata.getBoolean(
		    "com.cattailsw.nanidroid.DEVICE_VALIDATION_NO_TELEMETRY", false);
	} catch (PackageManager.NameNotFoundException ignored) {
	    return false;
	}
    }

}

File: src/com/cattailsw/nanidroid/CatTailApplication.kt
package com.cattailsw.nanidroid

import android.app.Application
import android.content.pm.PackageManager
import com.cattailsw.nanidroid.util.AnalyticsUtils
import com.cattailsw.nanidroid.util.CrashReporting

class CatTailApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        if (isDeviceValidationNoTelemetry()) {
            AnalyticsUtils.setDeviceValidationNoTelemetry(true)
            return
        }
        CrashReporting.initialize(this)
    }

    private fun isDeviceValidationNoTelemetry(): Boolean = try {
        val info = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA)
        info.metaData?.getBoolean(DEVICE_VALIDATION_NO_TELEMETRY, false) == true
    } catch (_: PackageManager.NameNotFoundException) {
        false
    }

    private companion object {
        const val DEVICE_VALIDATION_NO_TELEMETRY =
            "com.cattailsw.nanidroid.DEVICE_VALIDATION_NO_TELEMETRY"
    }
}
